### PR TITLE
[Inductor] Remove VecChecker and fallback non-supported Vec op to Scalar impl with a for loop

### PR DESCRIPTION
**Pull request by zhuhaozhe**
Fall back non-vectorized op by scalar impl + for loop.

Example code:
```
cpp_fused_igammac_0 = async_compile.cpp_pybinding(['const double*', 'const double*', 'double*'], '''
#include "/tmp/torchinductor_root/z4/cz4j2mmotlx3z2b7u4fbjtdt4x6plhd67ljwzg5bk7ekv4xz6y7q.h"
extern "C"  void kernel(const double* in_ptr0,
                       const double* in_ptr1,
                       double* out_ptr0)
{
    {
        for(int64_t x0=static_cast<int64_t>(0L); x0<static_cast<int64_t>(48L); x0+=static_cast<int64_t>(8L))
        {
            auto tmp0 = at::vec::VectorizedN<double,2>::loadu(in_ptr0 + static_cast<int64_t>(x0), 8);
            auto tmp1 = in_ptr1[static_cast<int64_t>(0L)];
            auto tmp2 = at::vec::VectorizedN<double,2>(tmp1);
            auto tmp3 =
            [&]()
            {
                __at_align__ std::array<double, 8> tmpbuf0;
                tmp0.store(tmpbuf0.data(), 8);
                __at_align__ std::array<double, 8> tmpbuf1;
                tmp2.store(tmpbuf1.data(), 8);
                __at_align__ std::array<double, 8> tmpbuf_out;
                for (int i = 0; i < 8; i++)
                {
                    tmpbuf_out[i] = calc_igammac(tmpbuf0[i], tmpbuf1[i]);
                }
                return at::vec::VectorizedN<double, 2>::loadu(tmpbuf_out.data(), 8);
            }
            ()
            ;
            tmp3.store(out_ptr0 + static_cast<int64_t>(x0), 8);
        }
        #pragma omp simd simdlen(4) 
        for(int64_t x0=static_cast<int64_t>(48L); x0<static_cast<int64_t>(50L); x0+=static_cast<int64_t>(1L))
        {
            auto tmp0 = in_ptr0[static_cast<int64_t>(x0)];
            auto tmp1 = in_ptr1[static_cast<int64_t>(0L)];
            auto tmp2 = calc_igammac(tmp0, tmp1);
            out_ptr0[static_cast<int64_t>(x0)] = tmp2;
        }
    }
}
''')

```

`frexp` are difficult to be handled by common `fallback` since it returns two `cse_var` https://github.com/pytorch/pytorch/blob/2ba60a1618813c91b26ca0860242626b22832295/torch/_inductor/codegen/cpp.py#L752-L766
So we added a special function to do that.
```
cpp_fused_frexp_0 = async_compile.cpp_pybinding(['const double*', 'double*', 'int32_t*'], '''
#include "/tmp/torchinductor_root/z4/cz4j2mmotlx3z2b7u4fbjtdt4x6plhd67ljwzg5bk7ekv4xz6y7q.h"
extern "C"  void kernel(const double* in_ptr0,
                       double* out_ptr0,
                       int32_t* out_ptr1)
{
    {
        for(int64_t x0=static_cast<int64_t>(0L); x0<static_cast<int64_t>(16L); x0+=static_cast<int64_t>(8L))
        {
            auto tmp0 = at::vec::VectorizedN<double,2>::loadu(in_ptr0 + static_cast<int64_t>(x0), 8);
            at::vec::Vectorized<int32_t> tmp1;
            at::vec::VectorizedN<double, 2> tmp2;
            [&]()
            {
                __at_align__ std::array<double, 8> tmpbuf;
                tmp0.store(tmpbuf.data(), 8);
                __at_align__ std::array<int32_t, 8> tmpbuf_exponent;
                __at_align__ std::array<double, 8> tmpbuf_mantissa;
                for (int i = 0; i < 8; i++)
                {
                    tmpbuf_mantissa[i] = std::frexp(tmpbuf[i], &tmpbuf_exponent[i]);
                }
                tmp1 = at::vec::Vectorized<int32_t>::loadu(tmpbuf_exponent.data(), 8);
                tmp2 = at::vec::VectorizedN<double, 2>::loadu(tmpbuf_mantissa.data(), 8);
            }
            ();
            tmp2.store(out_ptr0 + static_cast<int64_t>(x0), 8);
            tmp1.store(out_ptr1 + static_cast<int64_t>(x0), 8);
        }
        #pragma omp simd simdlen(4) 
        for(int64_t x0=static_cast<int64_t>(16L); x0<static_cast<int64_t>(20L); x0+=static_cast<int64_t>(1L))
        {
            auto tmp0 = in_ptr0[static_cast<int64_t>(x0)];
            int32_t tmp1;
            auto tmp2 = std::frexp(tmp0, &tmp1);
            out_ptr0[static_cast<int64_t>(x0)] = tmp2;
            out_ptr1[static_cast<int64_t>(x0)] = tmp1;
        }
    }
}
''')
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134569

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by jgong5** [comment]
@zhuhaozhe could you please also check if these ops can only be implemented with scalarization or it is still possible to vectorize them?

**Comment by zhuhaozhe** [comment]
> @zhuhaozhe could you please also check if these ops can only be implemented with scalarization or it is still possible to vectorize them?

@jgong5, sure.
For this PR, I can think we can just implemented with scalarization since those ops are not exposed by dynamo benchmark and only exist in UT. We should not have performance regression here and we can enhance it with real vectorization function one by one later.

Current ops in `CppOverride` while not in `CppVecOverride`.:
image

File: torch/_inductor/codegen/cpp.py [hunk 1]
```diff
@@ -3486,6 +3365,41 @@ def _select_tiling_indices(
             return contig_vars_sorted
         return sorted(contig_vars_sorted, key=contig_vars_list.count)[-1:]
 
+    @staticmethod
+    def _dtype_not_supported(loop_bodies, supported_dtype):
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:3374]: Is it better to factor out a function that returns a set of all the used dtypes in the loop bodies and check the supported dtypes on the returned value? Then, we can perhaps replace `get_loop_body_lowp_fp` with it too?

File: torch/_inductor/codegen/cpp.py [hunk 2]
```diff
@@ -3744,29 +3662,15 @@ def run(kernel):
...
-                        could_masked_vec = (
-                            could_masked_vec and vec_checker.simd_masked_vec
-                        )
-                        if not could_vec:
-                            tiling_factors = []
-                            tiling_indices = []
-                            break
+            if TilingSelect._dtype_not_supported(
+                TilingSelect._get_loop_body(fn_list), MASKED_VECTORIZABLE_DTYPES
+            ):
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:3673]: Why do we need to check it again?

File: torch/_inductor/codegen/cpp.py [hunk 3]
```diff
@@ -925,7 +942,55 @@ class CppVecOverrides(CppOverrides):
     def __new__(cls, *args, **kargs):
         self = super().__new__(cls)
 
-        def wrap(func):
+        def vectorize_by_scalar_func_with_for_loop(scalar_func, *args, **kwargs):
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:945]:
```suggestion
        def scalarize(scalar_func, *args, **kwargs):
```

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/codegen/cpp.py (hunk 1 above)
Comment (zhuhaozhe) [reply torch/_inductor/codegen/cpp.py:3374]: Hi, @jgong5 . From my understanding, `get_loop_body_lowp_fp` is not only check the dtype but also check  whether the `_node.target` is in `[load, store, bas, neg, output]`.  So maybe we can leave them as two util fuctions?

File: torch/_inductor/codegen/cpp.py (hunk 2 above)
Comment (zhuhaozhe) [reply torch/_inductor/codegen/cpp.py:3673]: This is check whether we can use masked_vectorize. After https://github.com/pytorch/pytorch/pull/128802 merged, we can simply remove this check.

File: torch/_inductor/codegen/cpp.py (hunk 3 above)
Comment (zhuhaozhe) [reply torch/_inductor/codegen/cpp.py]: Thanks for the advice, updated!

**Comment by zhuhaozhe** [comment]
The model level CI fail caused by network issue, should be irrelevant
image

File: torch/_inductor/codegen/cpp.py [hunk 4]
```diff
@@ -3486,6 +3374,41 @@ def _select_tiling_indices(
             return contig_vars_sorted
         return sorted(contig_vars_sorted, key=contig_vars_list.count)[-1:]
 
+    @staticmethod
+    def _get_dtype_from_loopbodies(loop_bodies):
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:3378]: How about moving it to cpp_utils.py, considering it is used by other classes too?

File: torch/_inductor/codegen/cpp.py [hunk 5]
```diff
@@ -3486,6 +3374,41 @@ def _select_tiling_indices(
...
+            ]
+            for graph in graphs:
+                for node in graph.nodes:
+                    if node.op != "call_method":
+                        continue
+                    dtypes.add(node.meta[OptimizationContext.key].dtype)
+        return dtypes
+
+    @staticmethod
+    def _get_loop_body(fn_list):
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:3392]: ditto

**Review by jgong5 — APPROVED** [review]
LGTM. Some minor comment on code movement.

File: torch/_inductor/codegen/cpp.py (hunk 4 above)
Comment (zhuhaozhe) [reply torch/_inductor/codegen/cpp.py]: Updated, thanks for the suggestion.

File: torch/_inductor/codegen/cpp.py (hunk 5 above)
Comment (zhuhaozhe) [reply torch/_inductor/codegen/cpp.py]: updated

**Review by jansel — APPROVED** [review]

**Comment by ezyang** [comment]
This PR is responsible for -7.2% improvement in instruction count on basic_modules_ListOfLinears_inductor instruction count benchmark. Thanks for the improvement!

cc @laithsakka

**Comment by jgong5** [comment]
> This PR is responsible for -7.2% improvement in instruction count on basic_modules_ListOfLinears_inductor instruction count benchmark. Thanks for the improvement!
> 
> cc @laithsakka

We intended to simplify the cpp vec codegen, not expecting instruction count reduction indeed... but good not know it helped...

**Pull request closed without merging**